File: heavy-metal-bake-sale/csharp/src/HeavyMetalBakeSale/Product.cs
namespace HeavyMetalBakeSale;

public class Product
{
    public string Name { get; }
    public Money Price { get; }
    public string PurchaseCode { get; }
    public int Stock { get; private set; }

    public Product(string name, Money price, string purchaseCode, int stock)
    {
        Name = name;
        Price = price;
        PurchaseCode = purchaseCode;
        Stock = stock;
    }

    public bool IsInStock => Stock > 0;

    public void DecrementStock()
    {
        Stock--;
    }
}

File: heavy-metal-bake-sale/typescript/src/Product.ts
import { Money } from './Money.js';

export class Product {
  private _stock: number;

  constructor(
    public readonly name: string,
    public readonly price: Money,
    public readonly purchaseCode: string,
    stock: number,
  ) {
    this._stock = stock;
  }

  get stock(): number {
    return this._stock;
  }

  get isInStock(): boolean {
    return this._stock > 0;
  }

  decrementStock(): void {
    this._stock--;
  }
}
